# Soma de Riemann

A forma mais facil para integrar uma função é calcular a soma de Riemann. Aprendemos sobre essa soma no calculo II para definir uma integral definida. 

$$\int_a^b f(x)dx$$

Nessa tecnica, dividimos o intervalo $[a,b]$ em $n$ intervalo pequeno, aí a largura de cada intervalo pequeno é 

$$\Delta x = \dfrac{b - a}{n}$$

Para cada intervalo pequeno calculamos o valor da função em um ponto do intervalo. Depois, para cada intervalo, temos $\Delta x \times f(x_i)$, ou seja, area de um retangulo. Somando a area de todos os retangulos pequeno obtemos o valor aproximado da integral. No limite de $\Delta x \rightarrow 0$, a aproximação tende ao valor real. 

<img src="./images/soma.jpg" style="width:300px;height:350px;">

Agora temos uma questão em aberta. Qual ponto do intervalo deve ser usando para calcular $f(x_i)$? Aqui tem tres opções comuns: 1) usar o ponto inicial do intervalo, 2) ponto do meio, 3) ponto final. 


In [2]:
import numpy as np
def riemann_sum(func,a,b,N,method='midpoint'):
    '''Compute the Riemann sum of f(x) over the interval [a,b].
    
    Credit to: https://www.math.ubc.ca/~pwalls/math-python/integration/riemann-sums/
    
    Parameters
    ----------
    `func` : function
        Vectorized function of one variable
    `a` , `b` : numbers
        Endpoints of the interval [a,b]
    `N` : integer
        Number of subintervals of equal length in the partition of [a,b]
    `method` : string
        Determines the kind of Riemann sum:
        `right` : Riemann sum using right endpoints
        `left` : Riemann sum using left endpoints
        `midpoint` (default) : Riemann sum using midpoints

    Returns
    -------
    float
        Approximation of the integral given by the Riemann sum.
    '''
    a, b = float(a), float(b)
    dx = (b - a)/N
    x = np.linspace(a,b,N+1)
    
    var = list(func.free_symbols)[0]
    f_np = sp.lambdify(var, func)

    if method == 'left':
        x_left = x[:-1]
        return np.sum(f_np(x_left)*dx)
    elif method == 'right':
        x_right = x[1:]
        return np.sum(f_np(x_right)*dx)
    elif method == 'midpoint':
        x_mid = (x[:-1] + x[1:])/2
        return np.sum(f_np(x_mid)*dx)
    else:
        raise ValueError("Method must be 'left', 'right' or 'midpoint'.")

## **Exemplo:** 
Calcular o valor da integral 

$$\int_0^2 \cos(x) + 2 $$

usando a soma de Riemann e comprar com o valor real

**Resolução:**

In [10]:
import sympy as sp
x = sp.symbols('x')

In [20]:
func = lambda x: sp.cos(x) + 2

In [21]:
func(x)

cos(x) + 2

In [27]:
sp.integrate(func(x),( x, 0, 2))
sp.N(_)

4.91081469248160

In [26]:
riemann_sum(func(x), 0,2, 10)

4.910814692481599

In [28]:
riemann_sum(func(x), 0,2, 10, method='right')

4.764649729160901

In [29]:
riemann_sum(func(x), 0,2, 10, method='left')

5.04787909647033

# Valor medio de uma função continua

A média de $n$ pontos discretos pode ser calculada usando a seguinte formula

$$\text{Media} = \dfrac{\sum_{i=1}^n y_i}{n} $$

em que $y_i$ são medidas individuais. 

Em contraste, suponha que y seja uma função no contínuo de uma variável independente $x$. Nesse caso, existe um número infinito de valores entre $a$ e $b$. Do mesmo modo que a equação acima pode ser aplicada para determinar a média de leituras discretas, você poderia também estar interessado em calcular a média de uma função no contínuo $y = f(x)$ no intervalo de $a$ até $b$. A integração é usada para esse propósito, como especificado pela fórmula

$$\text{Media} = \dfrac{\int_a^b f(x)dx}{b - a} $$

Essa fórmula tem centenas de aplicações em engenharia. Por exemplo, é usada para calcular o centro de gravidade de objetos irregulares na mecânica e na engenharia civil e para determinar a corrente eficaz na engenharia elétrica. Até a densidade pode ser considerado um caso dessa média. 

# Fórmulas de Integração de Newton-Cotes

As fórmulas de Newton-Cotes são os esquemas mais comuns de integração numérica.
Elas são baseadas na estratégia de substituir uma função complicada ou dados tabulados
por uma função aproximadora simples que seja fácil de integrar:

$$I = \int_a^b f(x)dx \approx \int_a^b f_n(x)dx $$

em que $f_n(x)$ é um polinômio da forma

$$f_n(x) = a_0 + a_1 x + ...+ a_{n-1}x^{n-1} + a_n x^n $$

em que $n$ é o grau do polinômio.

Por exemplo, na figura abaixo (a), um polinômio de primeiro grau, $n = 1$, (uma reta) é usado como uma aproximação. Na (b), é usada uma parábola, $n = 2$, para o mesmo propósito.

<img src="./images/cotes.jpg" style="width:600px;height:350px;">

A soma de Riemann pode ser considerado um caso especifico da formula de Newton-Cotes com $n = 0$, uma linha constante. 


## A Regra do Trapézio

A regra do trapézio é a primeira fórmula de integração fechada de Newton-Cotes. Ela
corresponde ao caso no qual o polinômio na equação geral é de primeiro grau:

$$I = \int_a^b f(x)dx \approx \int_a^b f_1(x)dx $$

Lembre-se que a equação de uma reta que passa pelos pontos $(a, f(a))$ e $(b, f(b))$ é 

$$f_1(x) = f(a) + \dfrac{f(b) - f(a)}{b - a} (x - a)$$

Inserindo essa equação na integral chegamos na formula de *regra do trapézio*:

$$I = (b -a) \frac{f(b) + f(a)}{2} $$

Geometricamente, a regra dos trapézios é equivalente a aproximar a integral pela
área do trapézio sob a reta ligando $f(a)$ e $f(b)$

<img src="./images/trapezio.jpg" style="width:400px;height:350px;">

## Erro na Regra do Trapézio

Quando empregamos a integral sob um segmento de reta para aproximar a integral sob a
curva, obviamente incorremos em um erro que pode ser substancial. Uma estimativa para o erro de truncamento local de uma única aplicação da regra do trapézio é 

$$E_t = - \dfrac{1}{12} f''(\xi)(b - a)^3 $$

em que $\xi$ está em algum ponto do intervalo entre $a$ e $b$.

## **Exemplo:**

Use a regra do trapézio para calcular a integral numericamente 

$$
f(x) = 0.2 + 25 x - 200 x^2 + 675 x^3 - 900 x^4 + 400 x^5
$$

de $a = 0$ a $b = 0.8$.

**Resolução:**

Antes de resolver essa integral numericamente vamos resolver-la analiticamente

In [30]:
f = lambda x: 0.2 + 25*x - 200 * x**2 + 675 * x**3 - 900 * x**4 + 400 * x**5
f(x)

400*x**5 - 900*x**4 + 675*x**3 - 200*x**2 + 25*x + 0.2

In [32]:
sp.integrate(f(x), (x, 0, 0.8))

1.64053333333333

A solução numerica usando a regra do trapézio é 

In [35]:
a = 0
b = 0.8

I = (b - a)*(f(x).subs(x,a) + f(x).subs(x,b))/(2)
I

0.172800000000011

Para calcular o erro podemos usar a media da $f''(x)$ no intervalo em vez de $f''(\xi)$

In [37]:
f2 = sp.integrate(f(x).diff(x,2), (x, 0, 0.8))/(b - a)
f2

-60.0000000000003

In [38]:
E_a = -1/12 * (f2)*(b - a)**3

E_a

2.56000000000001

### Aplicação Múltipla da Regra de Trapézio

Uma maneira de melhorar a acurácia da regra do trapézio é dividir o intervalo de integração de a a b em diversos segmentos e aplicar o método a cada segmento

<img src="./images/trapezio2.jpg" style="width:250px;height:800px;">

As áreas correspondentes aos segmentos individuais podem então ser somadas para fornecer a integral para o intervalo inteiro. As equações resultantes são chamadas *fórmulas de integração por aplicações múltiplas* ou *compostas*.

imagine que existem $n + 1$ pontos base igualmente espaçados $(x_0, x_1, x_2, ..., x_n)$. Conseqüentemente, existem $n$ segmentos de largura igual:

$$h = \frac{b - a}{n} $$

Se $a$ e $b$ forem designados por $x_0$ e $x_n$, respectivamente, a integral total pode ser representada como

$$
I = \int_{x_0}^{x_1} f(x) dx + \int_{x_1}^{x_2} f(x) dx + ... +\int_{x_{n-1}}^{x_n} f(x) dx
$$

Substituindo cada integral pela regra do trapézio, obtém-se

$$
I = h \frac{f(x_0) + f(x_1)}{2} + h \frac{f(x_1) + f(x_2)}{2} + ... + h \frac{f(x_{n-1}) + f(x_n)}{2}
$$

ou, agrupando termos,

$$I = \frac{h}{2} \big[f(x_0) + 2 \sum_{i=1}^{n-1} f(x_i) + f(x_n) \big]$$

Um erro para a aplicação múltipla da regra do trapézio pode ser obtido pela soma dos
erros individuais em cada segmento, o que dá

$$E_t = - \dfrac{(b-a)^3}{12 n^3} \sum_{i=1}^n f''(\xi) $$

em que $f''(\xi_i)$ é a segunda derivada em um ponto $\xi_i$ localizado no segmento $i$. Esse resultado pode ser simplificado por uma estimativa do valor médio da segunda derivada no intervalo todo como

$$\bar{f''} \approx \dfrac{\sum_{i=1}^n f''(\xi_i)}{n} $$

Usando essa aproximação, $\sum f''(\xi_i) \approx n \bar{f''}$ , a formula de erro fica

$$E_t = - \dfrac{(b-a)^3}{12 n^2} \bar{f''}$$

Logo, se o número de segmentos for dobrado, o erro de truncamento será dividido por
quatro.

### **Exemplo:**
 Use a regra do trapézio com dois e quatro segmentos para obter uma estimativa da integral de

$$
f(x) = 0.2 + 25 x - 200 x^2 + 675 x^3 - 900 x^4 + 400 x^5
$$

de $a = 0$ a $b = 0.8$.

**Resolução:**

Para $n=2 (h = 0.4) $ temos

In [58]:
def trapezoidal(func, a, b, n=1):
    assert a < b, 'the inferior limit must be less than superior one'
    h = (b - a)/n
    xi = np.linspace(a, b, n+1)
    return h/2*(func(xi[0]) + 2 * sum(map(func, xi[1:-1])) + func(xi[-1]))

In [47]:
list(map(lambda x: x**2, [ 1,2,3]))

[1, 4, 9]

In [48]:
sum(map(lambda x: x**2, [ 1,2,3]))

14

In [65]:
f = lambda x: 0.2 + 25*x - 200 * x**2 + 675 * x**3 - 900 * x**4 + 400 * x**5

for i in range(1, 10):
    print(f"n = {i}: {trapezoidal(f, 0, 0.8, i):.4f}")


n = 1: 0.1728
n = 2: 1.0688
n = 3: 1.3696
n = 4: 1.4848
n = 5: 1.5399
n = 6: 1.5703
n = 7: 1.5887
n = 8: 1.6008
n = 9: 1.6091


Tres observações sobre a regra do trapézio

- Para aplicações individuais em funções bem comportadas, a aplicação múltipla da regra do trapézio é decididamente boa para obter o tipo de acurácia necessária em muitas aplicações de engenharia.
- Se for necessária uma alta acurácia, a regra do trapézio com segmentos múltiplos exige um grande esforço computacional. Embora esse esforço possa ser desprezível para uma única aplicação, pode ser muito importante quando (a) muitas integrais estão sendo calculadas ou (b) a própria função leva muito tempo para ser calculada. Para tais casos, abordagens mais eficientes (como aquelas no restante deste capítulo e no próximo) podem ser necessárias.
- Finalmente, os erros de arredondamento podem limitar nossa habilidade de determinar integrais. Isso decorre tanto da precisão da máquina quanto da grande quantidade de cálculos envolvida em técnicas simples como a regra do trapézio com segmentos múltiplos.

## Regreas de Simpson

Além de aplicar a regra do trapézio com segmentos menores, outra forma de obter uma estimativa mais acurada de uma integral é usar polinômios de grau mais alto para ligar os pontos. Por exemplo, se existir um ponto extra no ponto médio entre $f(a)$ e $f(b)$, os três pontos
podem ser ligados por uma parábola.  Se existirem dois pontos igualmente espaçados entre $f(a)$ e $f(b)$, os quatro pontos podem ser ligados por um polinômio de $3º$ grau. As fórmulas que resultam de tomar as integrais desses polinômios são chamadas de *regras de Simpson*.

### A Regra $1/3$ de Simpson

A regra $1/3$ de Simpson é obtida quando um polinômio interpolador de segundo grau

$$I = \int_a^b f(x)dx \approx \int_a^b f_2(x)dx $$

Se $a$ e $b$ forem designados por $x_0$ e $x_2$ e se $f_2(x)$ for representado por um polinômio de Lagrange (lembre-se do capitulo sobre interpolação) de segundo grau, a integral se torna

$$
I = \int_{x_0}^{x_2} \big[\dfrac{(x - x_1)(x - x_2)}{(x_0 - x_1)(x_0 - x_2)}f(x_0) + \dfrac{(x - x_0)(x - x_2)}{(x_1 - x_0)(x_1 - x_2)}f(x_1) + \dfrac{(x - x_0)(x - x_1)}{(x_2 - x_0)(x_2 - x_1)}f(x_2) \big] dx
$$

Depois da integração e de manipulações algébricas, obtém-se a seguinte fórmula:

$$I \approx \dfrac{h}{3} \big[f(x_0) + 4f(x_1) + f(x_2) \big] $$

em que, para esse caso, $h = (b − a)/2$. Essa equação é conhecida como a *regra $1/3$ de
Simpson*. Ela é a segunda fórmula de integração fechada de Newton-Cotes. A designação
"1/3" vem do fato que $h$ está dividido por $3$ na equação.

É possível mostrar que a aplicação da regra $1/3$ de Simpson para um único segmento
tem um erro de truncamento de

$$E_t = - \frac{1}{90} h^5 f^{(4)}(\xi) $$

ou, como $h = (b − a)/2$,

$$E_t = - \frac{(b - a)^5}{2880} f^{(4)}(\xi) $$

em que $\xi$ é algum ponto no intervalo entre $a$ e $b$. Logo, a regra $1/3$ de Simpson é mais acurada do que a regra do trapézio. A comparação com a equação do erro da regra do trapézio indica que ela é mais acurada do que o esperado. Em vez de ser proporcional à terceira derivada, o erro é proporcional à quarta derivada.

## **Exemplo:**

Use a regra $1/3$ de Simpson para calcular a integral numericamente 

$$
f(x) = 0.2 + 25 x - 200 x^2 + 675 x^3 - 900 x^4 + 400 x^5
$$

de $a = 0$ a $b = 0.8$. Lembrando que o valor real é 1.640533

In [66]:
f = lambda x: 0.2 + 25*x - 200 * x**2 + 675 * x**3 - 900 * x**4 + 400 * x**5

In [68]:
a = 0
b = 0.8
h = (b - a)/2

I = h/3*(f(0) + 4*f(0.4) + f(0.8))
print(I)

1.3674666666666742


### Aplicações Múltiplas da Regra $1/3$ de Simpson

Do mesmo modo como no caso da regra do trapézio, a regra de Simpson pode ser melhorada dividindo-se o intervalo de integração em diversos segmentos de mesmo comprimento

$$h = \dfrac{b - a}{2} $$

A integral total pode ser representada como

$$
I = \int_{x_0}^{x_2} f(x) dx + \int_{x_2}^{x_4} f(x) dx + ... +\int_{x_{n-2}}^{x_n} f(x) dx
$$

Substituindo cada integral individual pela regra $1/3$ de Simpson, obtemos

$$
I = 2h \frac{f(x_0) + 4f(x_1) + f(x_2)}{6} + 2h \frac{f(x_2) + 4f(x_3) + f(x_4)}{6} + ... + 2h \frac{f(x_{n-2}) + 4f(x_{n-1}) +f(x_n)}{6}
$$

ou, combinando os termos

$$
I \approx \frac{b - a}{3n}\big[f(x_0) + 4\sum_{i=1,3,5,...}^{n-1}f(x_i) + 2\sum_{i=2,4,6,...}^{n-2}f(x_i) + f(x_n) \big]
$$

Uma estimativa de erro para a aplicação múltipla da regra de Simpson é obtida da mesma maneira que para a regra trapezoidal, somando-se os erros individuais para os segmentos e fazendo a média da derivada, resultando em

$$E_t = - \frac{(b-a)^5}{180 n^4} \bar{f}^{(4)} $$

em que $\bar{f}^{(4)} $ é o valor médio da quarta derivada no intervalo.

### **Exemplo:**
 Use a regra $1/3$ do Simpson com $n=4$ para obter uma estimativa da integral de

$$
f(x) = 0.2 + 25 x - 200 x^2 + 675 x^3 - 900 x^4 + 400 x^5
$$

de $a = 0$ a $b = 0.8$.

**Resolução:**



In [82]:
def simpson13(func, a, b, n=2):
    assert a < b, "The inferior limit must be less than the superior one"
    assert n > 1, "The simpson 1/3 needs at least 3 points, n = points - 1"
    assert n % 2 == 0, "The simpson 1/3 works only for par number of segments "

    xi = np.linspace(a, b, n+1)
    return ((b - a)/(3*n))*(func(xi[0]) + 4 * sum(map(func, xi[1:-1:2])) \
           + 2 * sum(map(func, xi[2:-2:2])) + func(xi[-1]))
    

In [83]:
simpson13(f, 0, 0.8, 4)

1.6234666666666717

O exemplo anterior ilustra que a versão com aplicações múltiplas da regra $1/3$ de
Simpson fornece resultados muito acurados. Por essa razão, ela é considerada superior à
regra do trapézio na maioria das aplicações. Entretanto, como mencionado anteriormente,
ela é limitada aos casos nos quais os valores estão igualmente espaçados. Além disso, é limitada a situações nas quais haja um número par de segmentos e um número ímpar de pontos. Conseqüentemente, como discutido na próxima seção, uma fórmula para um número ímpar de segmentos e par de pontos, conhecida como regra $3/8$ de Simpson, é usada em conjunto com a regra $1/3$ para permitir o cálculo tanto para um número par quanto para um número ímpar de segmentos.

## Regra $3/8$ de Simpson
De uma maneira parecida com a dedução da regra do trapézio e da regra $1/3$ de Simpson,
um polinômio de Lagrange de ordem tres pode ser ajustado a quatro pontos e integrado:

$$I = \int_a^b f(x)dx \approx \int_a^b f_3(x)dx $$
 
para fornecer

$$I \approx \frac{3h}{8} \big[f(x_0) + 3f(x_1) + 3f(x_2) + f(x_3)  \big] $$

em que $h = (b − a)/3$. Essa equação é chamada regra $3/8$ de Simpson porque $h$ é ultiplicada por $3/8$. Ela é a terceira fórmula de integração fechada de Newton-Cotes. A regra $3/8$ pode também ser expressa na forma

$$
I \approx \frac{b - a}{8}\big[f(x_0) + 3f(x_1) + 3f(x_2) + f(x_3) \big]
$$


Assim, os dois pontos interiores têm pesos de três oitavos, enquanto as extremidades têm
peso de um oitavo. A regra $3/8$ de Simpson tem um erro de

$$E_t = -\frac{3}{80} h^5 f^{(4)}(\xi) $$

ou, como $h = (b − a)/3$,

$$E_t = -\frac{(b-a)^5}{6480} f^{(4)}(\xi) $$

a regra $3/8$ é um pouco mais acurada do que a regra $1/3$.

A regra $1/3$ de Simpson é usualmente o método preferido, pois alcança uma acurácia de terceira ordem com três pontos em vez dos quatro pontos necessários para a versão $3/8$.
Entretanto, a regra $3/8$ tem utilidade quando o número de segmentos é ímpar. No caso do
ultimo exemplo, usamos a regra de Simpson para integrar a função com quatro segmentos.
Suponha que você quisesse uma estimativa usando cinco segmentos. Uma opção seria usar
a versão com aplicação múltipla da regra do trapézio. Entretanto, isso pode não ser aconselhável por causa do grande erro de truncamento associado com esse método. Uma alternativa seria aplicar a regra $1/3$ de Simpson aos primeiros dois segmentos e a regra $3/8$ de Simpson aos últimos três (figura abaixo). Dessa maneira, você poderia obter uma estimativa acurada até ordem três em todo o intervalo

<img src="./images/simpson.jpg" style="width:300px;height:400px;">


### **Exemplo:**

- (a) Use a regra 3/8 de Simpson para integrar

$$
f(x) = 0.2 + 25 x - 200 x^2 + 675 x^3 - 900 x^4 + 400 x^5
$$

de $a = 0$ a $b = 0.8$.

- (b) Use-a em conjunto com a regra 1/3 de Simpson para integrar a mesma função
usando cinco segmentos.

**Resolução:**

(a)

In [87]:
f = lambda x: 0.2 + 25*x - 200 * x**2 + 675 * x**3 - 900 * x**4 + 400 * x**5

a = 0
b = 0.8

In [86]:
xi = np.linspace(0, 0.8, 4)
xi

array([0.        , 0.26666667, 0.53333333, 0.8       ])

In [88]:
I1 = (b - a)/8 * (f(xi[0]) + 3 * (f(xi[1]) + f(xi[2])) + f(xi[-1]))
I1

1.5191703703703778

(b)

In [94]:
def simpson38(func, xi):
    assert len(xi) == 4,  "This version is designed only for 4 points integration"

    h = xi[1] - xi[0]
    return 3*h/8 *(f(xi[0]) + 3 * (f(xi[1]) + f(xi[2])) + f(xi[-1]))
    

In [102]:
xi = np.linspace(0, 0.8, 3+1)
xi

array([0.        , 0.26666667, 0.53333333, 0.8       ])

In [103]:
simpson38(f, xi)

1.5191703703703778

In [113]:
def integrate(func, a, b, n=2):
  assert a < b, 'the inferior limit must be less than the superior'
  
  xi = np.linspace(a, b, n+1)
  if n == 1:
    return trapezoidal(func, a, b)
  elif n % 2 == 0:
    return simpson13(func, a, b, n)
  elif n == 3:
    return simpson38(func, xi)
  else:
    h = (b - a)/n
    return simpson38(func, xi[-4:]) + simpson13(func, a, b-3*h, n-3 )


In [114]:
integrate(f, 0, 0.8, 5)

1.6450771626666787

## Integração com Segmentos Desiguais

Até esse ponto, todas as fórmulas para integração numérica foram baseadas em dados igualmente espaçados. Na prática, existem muitas situações nas quais essa hipótese não é válida e precisamos lidar com segmentos de tamanhos distintos. Por exemplo, dados obtidos experimentalmente, muitas vezes, são desse tipo. Para tais casos, um método é aplicar
a regra do trapézio para cada segmento e somar os resultados:

$$
I = h_1 \frac{f(x_0) + f(x_1)}{2} + h_2 \frac{f(x_1) + f(x_2)}{2} + ... + h_n \frac{f(x_{n-1}) + f(x_n)}{2}
$$

em que $h_i$ é a largura do segmento $i$. Observe que essa foi a mesma abordagem usada na
aplicação múltipla da regra do trapézio. A unica diferença é que nesse caso o valor do intervalo, $h_i$ não é igual para todos os intervalos. Portanto, não podemos simplificar essa equação como o caso de igualmente espaçado mas para o camputador é facil fazer esse calculo

### **Exemplo:**
A informação na tabela abaixo foi gerada usando o mesmo polinômio utilizado nos exemplos anteriores ($f(x) = 0,2 + 25x − 200x^2 + 675x^3 − 900x^4 + 400x^5$). Use a equação acima para determinar a integral a partir desses dados. Lembre que a resposta correta é 1,640533.

|$x$ | $f(x)$ |
|----|-------|
| 0  | 0.200   |
|  0.12 | 1.309729   |
| 0.22  | 1.305241   |
| 0.32  | 1.743393   |
|  0.36 | 2.074903   |
| 0.40  | 2.456000   |
| 0.44  | 2.842985   |
|  0.54 | 3.507297   |
|  0.64  |3.181929    |
|  0.70  |2.363000    |
|  0.80  |0.232000    |

**Resolução:**

$$
I = 0.12 \dfrac{1,309729 + 0,2}{2} + 0.10 \frac{1,305241 + 1,309729}{2} + ...+ 0.10 \frac{0,232 + 2,363}{2}\\
=  0,090584 + 0,130749 + · · · + 0,12975 = 1,594801
$$

o que representa um erro relativo porcentual absoluto de $\varepsilon_t = 2.8\%$.

Os dados do exemplo anterior estão mostrados na figura abaixo. Observe que alguns
segmentos adjacentes têm largura igual e, conseqüentemente, poderiam ter sido calculados usando as regras de Simpson. Isso usualmente leva a resultados mais precisos, como
ilustrado no exemplo a seguir.

<img src="./images/desigual.jpg" style="width:400px;height:350px;">

Neste caso, para melhorar a precisão, é melhor o algoritmo verifique se dois segmentos consecutivos tiverem comprimentos iguais, a regra 1/3 de Simpson é aplicada. Se três forem
iguais, a regra 3/8 é usada. Quando os segmentos adjacentes forem de comprimentos diferentes, a regra do trapézio é implementada.